### PR TITLE
HPCC-16095 Result Monitoring ECL code generator

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck anthony.fishbeck@lexisnexis.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@afishbeck Who should review this?

**Comment by afishbeck** [comment]
@rpastrana please review

File: esp/esdllib/esdl_def.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def.cpp]: perhaps outside of the scope of this issue, but all the available attributes should be well defined (#define, etc), here's a list of other attributes I found:
"logMethod"
"flReqFormat"
"flRespFormat"
"mbsiProductAssociations"
"item_tag"
"request_type"
"response_type"
"product_"
"productdefault_"
"productAssociation"
"ptdselectors"
"selectors"
"translations"
"optional"
"min_ver"
"max_ver"
"min_flver"
"max_flver"
"depr_flver"
"name"
"version"
"base_type"
"enum_type"
"type"
"ecl_hide"
"complex_type"
"variableRecordLengthProcessing"
"allowMultipleEntryPerUnitNumber"

File: esp/esdllib/esdl_def.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def.cpp]: def outside of the scope of this issue, but shouldn't jliball include jlib.hpp?

File: esp/esdllib/esdl_def.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def.cpp]: could "allTypes" be a member var?

File: esp/esdllib/esdl_def.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def.cpp]: second use of "allTypes" in the same class.

File: esp/esdllib/esdl_def_helper.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def_helper.cpp]: Out of curiosity, and not sure if this could ever happen but will getElements return attributes encapsulated in double quotes? "*[@ecl_hide=\"1"\]" ?

File: esp/esdllib/esdl_def_helper.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def_helper.cpp]: I'm sure this process is well though out but it's hard for me to visualize, since there's two loops within a parent loop. Let's just make sure this is the most efficient way to remove all hidden structs.

File: esp/esdllib/esdl_def_helper.cpp
Comment (rpastrana) [inline esp/esdllib/esdl_def_helper.cpp]: These removeXX methods seem duplicated in tools/esdlcmd/esdlcmd_diff.cpp, is it worthwhile to share?

File: esp/xslt/esdl2diffecl.xslt
Comment (rpastrana) [inline esp/xslt/esdl2diffecl.xslt]: minor, but date should be 2016 if new file

File: esp/xslt/esdl2diffecl.xslt
Comment (rpastrana) [inline esp/xslt/esdl2diffecl.xslt]: is this used?

File: esp/xslt/esdl2diffecl.xslt
Comment (rpastrana) [inline esp/xslt/esdl2diffecl.xslt]: The default value seems LN specific

File: esp/xslt/esdl2diffecl.xslt
Comment (rpastrana) [inline esp/xslt/esdl2diffecl.xslt]: can this be removed?

File: esp/xslt/esdl2diffecl.xslt
Comment (rpastrana) [inline esp/xslt/esdl2diffecl.xslt]: I could only skim over XSLT, it looked reasonable from a high level view, we might need an expert to review. BTW, noticed a few extra lines throughout, but didn't think it was worth pointing out inline.

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: date

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: These 3 default values look LN specific.

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: this looks intentional, but felt it was worth pointing out in case we didn't mean to hardcode an internal url

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: same xslt comment as above...

File: tools/esdlcmd/esdlcmd_common.hpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_common.hpp]: Good, this was needed.

File: tools/esdlcmd/esdlcmd_common.hpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_common.hpp]: This new var prevents us from editing path until we can confirm path is valid, but could we just do a path.clear in the fail case?

File: tools/esdlcmd/esdlcmd_diff.cpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_diff.cpp]: this looks to be static in nature, and prob useful for other ESDL commands, consider making it visible to other classes?

File: tools/esdlcmd/esdlcmd_diff.cpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_diff.cpp]: does it make sense to define this class, and some(if not all) of the helper methods below at the esdllib level?

**Comment by rpastrana** [comment]
@afishbeck made a few comments, please take a look.

File: esp/esdllib/esdl_def.cpp
Comment (afishbeck) [inline esp/esdllib/esdl_def.cpp]: That would be fine, but should be a new issue.  Here I just want to be consistent with existing code.

File: esp/esdllib/esdl_def.cpp
Comment (afishbeck) [inline esp/esdllib/esdl_def.cpp]: Probably, not sure why I "had" to add this.  I'll check.

File: esp/esdllib/esdl_def_helper.cpp
Comment (afishbeck) [inline esp/esdllib/esdl_def_helper.cpp]: Yes JLIB will treat both single and double quoted attributes equivalently.  I use single because the string is more readable without the escapes.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: I'll remove all these default values.  only made sense for my testing.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: again I'll remove

**Comment by afishbeck** [comment]
@rpastrana thanks I need to look at more closely or work on the other comments.

**Comment by rpastrana** [comment]
@afishbeck just sending the review back to you.

File: tools/esdlcmd/esdlcmd_diff.cpp
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_diff.cpp]: I may move this class somewhere even more common later... want to make it a bit more general though, after I've used it a bit more.

File: esp/esdllib/esdl_def.cpp
Comment (afishbeck) [inline esp/esdllib/esdl_def.cpp]: flags is a parameter to the method, not the class.. so allTypes has to be calculated locally.

File: esp/esdllib/esdl_def_helper.cpp
Comment (afishbeck) [inline esp/esdllib/esdl_def_helper.cpp]: I've commoned it up a bit, based on what level the processing is being done on.  I.e. structs or elements.

**Comment by afishbeck** [comment]
@rpastrana I've made extensive changes.  Some based on your comments, others based on external requirements and general refactoring.  Please re-review.

File: tools/esdlcmd/esdlcmd_monitor.cpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_monitor.cpp]: I'm surprised PropertyTree::queryPropTree accepts a StringBuffer, there's no need to call .str() on xpath?

File: tools/esdlcmd/esdlcmd_monitor.cpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_monitor.cpp]: Minor, but we should indent here for readability

**Review by rpastrana — APPROVED** [review]
@afishbeck just a few minor comments, otherwise looks good.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: For now I'd rather these be consistent for all monitors... that way I know what to expect.  I could make it part of the template later on.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: I left these in the generator for now because it allows me to quickly hand edit the generated code to make a debug query.  Don't want to have to regenerate.  Eventually I'll probably remove these entirely.

File: tools/esdlcmd/esdlcmd_common.cpp
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_common.cpp]: right, will change.

File: tools/esdlcmd/esdlcmd_monitor.cpp
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_monitor.cpp]: if the parameter is a char * it automatically casts.  If its a va_list (...) then you always have to use s.str().

File: tools/esdlcmd/esdlcmd_monitor.cpp
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_monitor.cpp]: definitely.  will do.

File: tools/esdlcmd/esdlcmd_monitor.cpp
Comment (rpastrana) [inline tools/esdlcmd/esdlcmd_monitor.cpp]: depTree->queryPropTree(xpath.setf("EsdlResponse[@name='%s']", resp_type.str()).str()); <- It seems like this last .str() is needed.

File: tools/esdlcmd/esdlcmd_common.cpp
Comment (afishbeck) [inline tools/esdlcmd/esdlcmd_common.cpp]: The FileIO destructor will handle that.

**Comment by afishbeck** [comment]
@rpastrana updated.

**Comment by afishbeck** [comment]
@rpastrana sorry but I also added another change.  Support for ESParray<string>.  Didn't want to wait for the pull request to be merged to add this.  Please take a look at the last 2 commits.

**Review by rpastrana — APPROVED** [review]
@afishbeck couple of comments

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: Minor: I would indent the children elements of <xsl:choose>

**Comment by rpastrana** [comment]
@afishbeck reviewed.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: Yes, this was actually removed in my last commit.  Good catch though.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: I'll have to deal with mac systems as a separate issue, I'm not set up to test at the moment.  It would only affect lines with certain characters.  Like this one, if it escapes the &apos;.

File: esp/xslt/esdl2monitor.xslt
Comment (afishbeck) [inline esp/xslt/esdl2monitor.xslt]: I'll fix this indent, but also rebase and squash, and ask Richard to pull.

File: esp/xslt/esdl2monitor.xslt
Comment (rpastrana) [inline esp/xslt/esdl2monitor.xslt]: I can make the mac on my desk available for test.

**Comment by afishbeck** [comment]
@rpastrana updated and rebased.

@richardkchapman this pull request is ready to merge.

**Comment by afishbeck** [comment]
Cool thanks, lets test tomorrow, and then if it has issues I'll open a JIRA
and fix it separately.

**Comment by afishbeck** [comment]
@rpastrana I was able to build on Mac once I figured out the right options.  But I didn't know how to install and run.  Instead I ran the xslt using both Xalan and libxslt on the mac and then made sure the resulting ECL compiled.  It worked so I don't think there are any issues.  The disable escaping option only needs to be used for certain output.

**Comment by afishbeck** [comment]
@richardkchapman this pull request is ready.

**Pull request merged**
